File: apps/api/src/application/services/couples.service.ts
import { Injectable } from '@nestjs/common';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import { CreateCoupleCommand, ConnectCoupleCommand } from '../commands';
import { GetCoupleQuery } from '../queries';
import { CoupleResponseDto, InviteCodeResponseDto } from '../dtos';
import {
  CoupleMatchingService,
  DDayCalculatorService,
} from '../../domain/services';

@Injectable()
export class CouplesService {
  constructor(
    private readonly commandBus: CommandBus,
    private readonly queryBus: QueryBus,
    private readonly coupleMatchingService: CoupleMatchingService,
    private readonly dDayCalculatorService: DDayCalculatorService,
  ) {}

  async createInvite(
    userId: string,
    startDate: Date,
  ): Promise<InviteCodeResponseDto> {
    await this.coupleMatchingService.validateCanCreateCouple(userId);

    const couple = await this.commandBus.execute(
      new CreateCoupleCommand(userId, startDate),
    );

    return {
      inviteCode: couple.inviteCode,
      expiresAt: couple.inviteCodeExpiresAt.toISOString(),
    };
  }

  async connectWithPartner(
    userId: string,
    inviteCode: string,
  ): Promise<CoupleResponseDto> {
    const couple = await this.commandBus.execute(
      new ConnectCoupleCommand(userId, inviteCode),
    );

    return CoupleResponseDto.fromEntity(couple);
  }

  async getCouple(
    userId: string,
    coupleId: string,
  ): Promise<CoupleResponseDto> {
    const result = await this.queryBus.execute(
      new GetCoupleQuery(userId, coupleId),
    );

    const dto = new CoupleResponseDto();
    dto.id = result.id;
    dto.user1Id = result.user1Id;
    dto.user2Id = result.user2Id;
    dto.startDate = result.startDate;
    dto.inviteCode = result.inviteCode;
    dto.inviteCodeExpiresAt = result.inviteCodeExpiresAt;
    dto.daysFromStart = this.dDayCalculatorService.calculateDaysFromStart(
      new Date(result.startDate),
    );
    dto.isComplete = result.isComplete;
    dto.createdAt = result.createdAt;
    dto.updatedAt = result.updatedAt;

    return dto;
  }
}

File: apps/api-kotlin/src/main/kotlin/com/couplecalendar/application/service/CouplesService.kt
package com.couplecalendar.application.service

import com.couplecalendar.application.command.couple.ConnectCoupleCommand
import com.couplecalendar.application.command.couple.ConnectCoupleCommandHandler
import com.couplecalendar.application.command.couple.CreateCoupleCommand
import com.couplecalendar.application.command.couple.CreateCoupleCommandHandler
import com.couplecalendar.application.command.couple.DisconnectCoupleCommand
import com.couplecalendar.application.command.couple.DisconnectCoupleCommandHandler
import com.couplecalendar.application.command.couple.UpdateCoupleStartDateCommand
import com.couplecalendar.application.command.couple.UpdateCoupleStartDateCommandHandler
import com.couplecalendar.application.dto.request.CreateCoupleRequest
import com.couplecalendar.application.dto.response.CoupleResponse
import com.couplecalendar.application.dto.response.InviteCodeResponse
import com.couplecalendar.application.query.couple.CoupleQueryResult
import com.couplecalendar.application.query.couple.GetCoupleQuery
import com.couplecalendar.application.query.couple.GetCoupleQueryHandler
import com.couplecalendar.application.query.couple.GetMyCoupleQuery
import com.couplecalendar.application.query.couple.GetMyCoupleQueryHandler
import com.couplecalendar.common.exception.BadRequestException
import com.couplecalendar.domain.service.DDayCalculatorService
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.format.DateTimeParseException
import java.util.UUID

@Service
class CouplesService(
    private val createCoupleCommandHandler: CreateCoupleCommandHandler,
    private val connectCoupleCommandHandler: ConnectCoupleCommandHandler,
    private val updateCoupleStartDateCommandHandler: UpdateCoupleStartDateCommandHandler,
    private val disconnectCoupleCommandHandler: DisconnectCoupleCommandHandler,
    private val getCoupleQueryHandler: GetCoupleQueryHandler,
    private val getMyCoupleQueryHandler: GetMyCoupleQueryHandler,
    private val dDayCalculatorService: DDayCalculatorService
) {

    fun createInvite(userId: UUID, request: CreateCoupleRequest): InviteCodeResponse {
        val startDate = parseDate(request.startDate)
        val command = CreateCoupleCommand(userId, startDate)
        val couple = createCoupleCommandHandler.handle(command)

        return InviteCodeResponse(
            inviteCode = couple.inviteCode!!,
            expiresAt = couple.inviteCodeExpiresAt!!.toString()
        )
    }

    fun connectWithPartner(userId: UUID, inviteCode: String): CoupleResponse {
        val command = ConnectCoupleCommand(userId, inviteCode)
        val couple = connectCoupleCommandHandler.handle(command)
        return CoupleResponse.fromAggregate(couple)
    }

    fun getCouple(userId: UUID, coupleId: UUID): CoupleResponse =
        getCoupleQueryHandler.handle(GetCoupleQuery(userId, coupleId)).toResponse()

    fun getMyCouple(userId: UUID): CoupleResponse =
        getMyCoupleQueryHandler.handle(GetMyCoupleQuery(userId)).toResponse()

    fun updateMyCoupleStartDate(userId: UUID, startDate: String): CoupleResponse {
        val command = UpdateCoupleStartDateCommand(userId, parseDate(startDate))
        val couple = updateCoupleStartDateCommandHandler.handle(command)
        return CoupleResponse.fromAggregate(couple)
    }

    fun disconnectMyCouple(userId: UUID) {
        disconnectCoupleCommandHandler.handle(DisconnectCoupleCommand(userId))
    }

    private fun parseDate(value: String): LocalDate =
        try {
            LocalDate.parse(value)
        } catch (e: DateTimeParseException) {
            throw BadRequestException("Invalid date format. Expected YYYY-MM-DD")
        }

    private fun CoupleQueryResult.toResponse(): CoupleResponse = CoupleResponse(
        id = id,
        user1Id = user1Id,
        user2Id = user2Id,
        startDate = startDate,
        inviteCode = inviteCode,
        inviteCodeExpiresAt = inviteCodeExpiresAt,
        daysFromStart = daysFromStart,
        isComplete = isComplete,
        createdAt = createdAt,
        updatedAt = updatedAt
    )
}
